src/device/socket/vsock.rs

```diff
@@ -398,28 +382,34 @@ impl<H: Hal, T: Transport> VirtIOSocket<H, T> {
         // buffer back to the RX queue then we don't access it again until next time it is popped.
         let header = unsafe {
             let buffer = Self::as_mut_sub_rx_buffer(self.rx_buf_dma.raw_slice(), token.into())?;
-            let _len = self.rx.pop_used(token, &[], &mut [buffer])?;
+            let packet_len = self.rx.pop_used(token, &[], &mut [buffer])?;
 
             // Read the header and body from the buffer. Don't check the result yet, because we need
             // to add the buffer back to the queue either way.
-            let header_result = read_header_and_body(buffer, body);
+            let header_result = read_header_and_body(buffer, body, packet_len);
 
             // Add the buffer back to the RX queue.
             let new_token = self.rx.add(&[], &mut [buffer])?;
             // If the RX buffer somehow gets assigned a different token, then our safety assumptions
             // are broken and we can't safely continue to do anything with the device.
-            assert_eq!(new_token, token);
-
+            if token != new_token {
+                return Err(SocketError::RecycledWrongBuffer.into());
```

Comment:
See the comment above. If we get the wrong token back here then our safety assumptions are broken and so anything more we do with the RX virtqueue buffers may be unsound, so we need to either panic or reset the device somehow, not just return an error and let the client keep using the device.